dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java

```diff
@@ -34,33 +36,119 @@
+class StaticLinkageChecker {
+
+  /**
+   * Given list of the jar files as file names in filesystem, outputs the report of static linkage
+   * check.
+   *
+   * @param arguments jar files available in the file system
+   * @throws IOException when there is a problem in reading a jar file
+   * @throws ClassNotFoundException when there is a problem in reading a class from a jar file
+   */
+  public static void main(String[] arguments) throws IOException, ClassNotFoundException {
+    String report = generateStaticLinkageReport(arguments);
+    System.out.println(report);
+  }
+
+  @VisibleForTesting
+  static String generateStaticLinkageReport(String[] jarFileNames) throws IOException,
+      ClassNotFoundException {
+    StringBuilder stringBuilder = new StringBuilder();
+    Set<FullyQualifiedMethodSignature> externalMethodReferences = new HashSet<>();
+    List<Path> paths = new ArrayList<>();
+    for (String jarFileName : jarFileNames) {
+      File jarFile = new File(jarFileName);
+      Path jarFilePath = jarFile.toPath();
+      paths.add(jarFilePath);
+      externalMethodReferences.addAll(listExternalMethodReferences(jarFilePath));
+    }
+    List<FullyQualifiedMethodSignature> unresolvedMethodReferences =
+        resolvedMethodReferences(paths, Lists.newArrayList(externalMethodReferences));
+    if (unresolvedMethodReferences.isEmpty()) {
+      stringBuilder.append("There were no unresolved method references from the jar file(s) :");
+      stringBuilder.append(paths);
+    } else {
+      stringBuilder.append("There were unresolved method references from the jar file(s):\n");
+      for (FullyQualifiedMethodSignature methodReference : unresolvedMethodReferences) {
+        stringBuilder.append("  ");
+        stringBuilder.append(methodReference);
+        stringBuilder.append("\n");
+      }
+    }
+    return stringBuilder.toString();
+  }
+
+  /**
+   * Checks the availability of the methods through the jar files, and lists the unavailable
+   * methods.
+   *
+   * @param paths jar files to search for the availability of the methods
+   * @param methodReferences methods to search for with the jar files
+   * @return list of methods that are not found in the jar files
+   */
+  static List<FullyQualifiedMethodSignature> resolvedMethodReferences(List<Path> paths,
+      List<FullyQualifiedMethodSignature> methodReferences) {
+    List<FullyQualifiedMethodSignature> unresolvedMethods = new ArrayList<>();
+
+    // Creates chain of ClassPath items in the same order as paths
+    ClassPath classPath = null;
+    for (Path path : paths) {
+      String pathFileName = path.toFile().getAbsolutePath();
+      classPath = new ClassPath(classPath, pathFileName);
+    }
+    SyntheticRepository repository = SyntheticRepository.getInstance(classPath);
+
+    for (FullyQualifiedMethodSignature methodReference : methodReferences) {
+      try {
+        JavaClass javaClass = repository.loadClass(methodReference.getClassName());
```

Comment:
Is `SyntheticRepository. loadClass` expensive? If so, this will be called way too often (once per reference instead of once per class).

Reply:
SyntheticRepository extends MemorySensitiveClassPathRepository.

`MemorySensitiveClassPathRepository.loadClass` has cache

> Find an already defined (cached) JavaClass object by name

https://commons.apache.org/proper/commons-bcel/apidocs/org/apache/bcel/util/MemorySensitiveClassPathRepository.html